public_html/wp-content/plugins/pattern-directory/bin/update-contains-block-types.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * Update all patterns with the "contains block types" meta field.
+ *
+ * To run locally, use `wp-env`, ex:
+ * yarn wp-env run cli "php wp-content/plugins/pattern-directory/bin/update-contains-block-types.php --all --per_page=100 --apply"
+ *
+ * To run in a sandbox, use php directly, ex:
+ * php ./bin/update-contains-block-types.php --all --apply
+ */
+
+namespace WordPressdotorg\Pattern_Directory;
+
+use const WordPressdotorg\Pattern_Directory\Pattern_Post_Type\{ POST_TYPE };
+
+// This script should only be called in a CLI environment.
+if ( 'cli' != php_sapi_name() ) {
+	die();
+}
+
+$opts = getopt( '', array( 'post:', 'url:', 'abspath:', 'per_page:', 'all', 'apply', 'verbose' ) );
+
+if ( empty( $opts['url'] ) ) {
+	$opts['url'] = 'https://wordpress.org/patterns/';
+}
+
+if ( empty( $opts['abspath'] ) && false !== strpos( __DIR__, 'wp-content' ) ) {
+	$opts['abspath'] = substr( __DIR__, 0, strpos( __DIR__, 'wp-content' ) );
+}
+
+$opts['apply']   = isset( $opts['apply'] );
+$opts['verbose'] = isset( $opts['verbose'] );
+$opts['all']     = isset( $opts['all'] );
+
+// Bootstrap WordPress
+$_SERVER['HTTP_HOST']   = parse_url( $opts['url'], PHP_URL_HOST );
+$_SERVER['REQUEST_URI'] = parse_url( $opts['url'], PHP_URL_PATH );
+
+require rtrim( $opts['abspath'], '/' ) . '/wp-load.php';
+
+if ( ! $opts['all'] && ! isset( $opts['post'] ) ) {
+	fwrite( STDERR, "Error! Either specify a post ID with --post=<ID> or explicitly run over --all.\n" );
+	die();
+}
+
+if ( ! $opts['apply'] ) {
+	echo "Dry run, will not update any patterns.\n";
+}
+
+$args = array(
+	'post_type'      => POST_TYPE,
+	'post_status'    => array( 'publish', 'pending' ),
+	'posts_per_page' => isset( $opts['per_page'] ) ? $opts['per_page'] : -1,
+	'post_parent'    => 0,
+	'orderby'        => 'date',
+	'order'          => 'DESC',
+	'meta_query'     => array(
+		// Only update patterns without this meta.
+		array(
+			'key'     => 'wpop_contains_block_types',
+			'compare' => 'NOT EXISTS',
+		),
+	),
+);
+if ( isset( $opts['post'] ) ) {
+	$args = array(
+		'post_type' => POST_TYPE,
+		'p' => absint( $opts['post'] ),
+	);
+}
+
+$query = new \WP_Query( $args );
+$meta_updated = 0;
+
+while ( $query->have_posts() ) {
+	$query->the_post();
+	$pattern    = get_post();
+	$pattern_id = $pattern->ID;
+	$blocks     = parse_blocks( $pattern->post_content );
+	$all_blocks = _flatten_blocks( $blocks );
+
+	// Get the list of block names and convert it to a single string.
+	$block_names = wp_list_pluck( $all_blocks, 'blockName' );
```

Comment:
I think it's better to remove the `core/` prefix from the blocks and the reason is that we can end up with a quite big `GET` request that could exceed the limit. If every core block is allowed it seems we have `60` of them.

If in the future we add support in pattern directory for third party blocks, we can revisit and I don't think a solution would be that hard.

For now in my GB PR, I'm stripping the `core/` prefix: https://github.com/WordPress/gutenberg/pull/45237/commits/0b3d6e088343db6f525197e8552507986fcf8a06

Reply:
> If every core block is allowed it seems we have 60 of them.

If every block is allowed, you could just not send the parameter, and we won't need to do any filtering.

> If in the future we add support in pattern directory for third party blocks

Yeah, this is trying to be future-friendly, so the full block name is necessary in the meta (otherwise we'll need another script to update it later…).

> we can end up with a quite big GET request that could exceed the limit. 

I think we're safe for now with just core blocks - are you running into the limit in your PR?

Reply:
>If every block is allowed, you could just not send the parameter, and we won't need to do any filtering.

That would be more convoluted as we would need to have operations in sync with the Blocks API(like `parent, ancestor, etc..`), so for now let's not do that. Even with having the above though, he problem remains if for example we have just one `disallowed` block. 

>I think we're safe for now with just core blocks - are you running into the limit in your PR?

With the prefix the length of `allowed_blocks` is `962` characters. It's quite long but I think we will be in the limit for now at least. Let's try that.

>Yeah, this is trying to be future-friendly, so the full block name is necessary in the meta (otherwise we'll need another script to update it later…).

We could just modify the script to add the prefix in `non-core` blocks, but as I said above we can try with the prefix for now and let's see if we need something different in the future.

Reply:
>That would be more convoluted as we would need to have operations in sync with the Blocks API(like parent, ancestor, etc..), so for now let's not do that.

Hm... I just realized that there are some nuances in this. What I'm sending you right now doesn't include the blocks for example that have a specific parent(ex `List item and List`), but your parsing of the content will probably have the `List item` as well, correct?

Reply:
> your parsing of the content will probably have the List item as well, correct?

Yes, it collapses down all the blocks used in a pattern, for example: `core/column,core/columns,core/group,core/heading,core/list,core/list-item,core/paragraph,core/separator,core/social-link,core/social-links`

Reply:
I have added the logic to include the parent/ancestor relationship and if all available core blocks are allowed to be inserted, I'm sending the `all` value as a `string`.

Reply:
> If every core block is allowed it seems we have 60 of them.

This is going to affect the cacheability of the API significantly if every API request sends a list of their installed blocks that we then filter by. Due to the Pattern APIs using WordPress to serve the API requests with nginx caching in-front, this could be a longer-term performance issue.

We should require/suggest strongly, at a minimum, that the allowed_blocks parameter should be sorted consistently to avoid two sites with different block orders getting different cache keys.

I can't see that clients could reasonably send a "I don't support this block" flag though, as it doesn't know what it doesn't know about.